src/Compilers/CSharp/Portable/Symbols/UsedAssembliesRecorder.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed class UsedAssembliesRecorder : BoundTreeWalkerWithStackGuard
+    {
+        private readonly CSharpCompilation _compilation;
+
+        public static void RecordUsedAssemblies(CSharpCompilation compilation, BoundNode node, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(node != null);
+
+            try
+            {
+                var visitor = new UsedAssembliesRecorder(compilation);
```

Comment:
>new UsedAssembliesRecorder(compilation) [](start = 30, length = 39)

It feels like it would be valuable to cache one `UsedAssembliesRecorder` per compilation (e.g. in the compilation).

Reply:
> It feels like it would be valuable to cache one UsedAssembliesRecorder per compilation (e.g. in the compilation).

The base class has state which cannot be shared while walking nodes in parallel.

---
In reply to: [346980102](https://github.com/dotnet/roslyn/pull/39807#discussion_r346980102) [](ancestors = 346980102)